### PR TITLE
RA-1899 pt 1: Write tests for latestobsforconceptlist

### PR DESCRIPTION
**Pull request by brandones**
https://issues.openmrs.org/browse/RA-1899

This also improves some UI text in the latestobsforconceptlist widget.

Screenshot from 2021-03-04 16-03-04

### CODE REVIEW

File: omod/src/main/web/dashboardwidgets/latestobsforconceptlist/latestobsforconceptlist.controller.js [hunk 1]
```diff
@@ -10,7 +10,7 @@ export default class LatestObsForConceptListController {
         this.maxAgeInDays = this.widgetsCommons.maxAgeToDays(this.config.maxAge);
 
         // Fetch last obs or obsGroup for the list of concepts
-        this.openmrsRest.list('latestobs', {
+        return this.openmrsRest.list('latestobs', {
```
Comment (brandones) [inline omod/src/main/web/dashboardwidgets/latestobsforconceptlist/latestobsforconceptlist.controller.js:13]: Returning the promise just allows us to run the assertions in the `then` block.

**Comment by brandones** [comment]
I'd recommend reviewing with "Hide whitespace."

**Comment by mogoodrich** [comment]
Will review in a bit, but also flagged @dkayiwa ... Daniel, anyone else you think would be good to review this?

File: package.json [hunk 2]
```diff
@@ -73,7 +73,7 @@
     "clean": "rimraf %npm_package_config_targetDir% && rimraf $npm_package_config_targetDir",
     "build:dev": "webpack --progress --colors --env=dev",
     "build:prod": "webpack --progress --colors --env=prod",
-    "build": "rimraf %npm_package_config_targetDir% && rimraf $npm_package_config_targetDir && karma start && webpack --progress --colors --env=prod",
+    "build": "rimraf %npm_package_config_targetDir% && rimraf $npm_package_config_targetDir && webpack --progress --colors --env=prod",
```
Comment (mogoodrich) [inline package.json:76]: just for reference, what is the change here?  karma is not needed/use?

File: omod/src/main/web/dashboardwidgets/dashboardwidgetscommons.service.js [hunk 3]
```diff
@@ -83,7 +83,7 @@ export default class WidgetsCommons {
             today.setMonth(today.getMonth()-parseInt(maxAge));
         } else if( maxAge.indexOf('y') !== -1 ){
             maxAge = maxAge.replace('y', '');
-            today.setDate(today.getYear()-(parseInt(maxAge)));
```
Comment (mogoodrich) [inline omod/src/main/web/dashboardwidgets/dashboardwidgetscommons.service.js:86]: Was this something that was wrong that was discovered during testing? (Just curious)

**Review by mogoodrich — APPROVED** [review]
Made a few comments/questions, but LGTM, thanks @brandones ... especially for taking the time to write all these much-needed tests.

In the PIH ticket, can you make sure to add to your notes for testing to test that the other things you changed still work as well (ie the display of the date to days message)?

File: package.json (hunk 2 above)
Comment (brandones) [reply package.json:76]: I don't think it makes sense to run the tests as part of building. This `build` command gets executed when you run the server. If you're developing, you want the server to start up regardless of whether or not the tests are currently green.

File: omod/src/main/web/dashboardwidgets/dashboardwidgetscommons.service.js (hunk 3 above)
Comment (brandones) [reply omod/src/main/web/dashboardwidgets/dashboardwidgetscommons.service.js:86]: Nah, it came up when I was trying things out on a server. I configured a widget using max age `1y` and got unexpected results.

File: package.json (hunk 2 above)
Comment (mogoodrich) [reply package.json:76]: got it... lets just confirm that this doesn't change the fact that they should be run during a a mvn build?

Comment (brandones) [reply package.json:76]: What's a mvn build? These are the lifecycle steps I see.

Screenshot from 2021-03-05 11-38-32

Comment (brandones) [reply package.json:76]: Shouldn't it run during test, not build? I would expect build to build. And test to test.

Comment (mogoodrich) [reply package.json:76]: sorry, it run if you a "mvn clean package"  it should run the tests... triggering that now, I see this:

[INFO] --- frontend-maven-plugin:1.4:npm (npm install) @ coreapps-omod ---
[INFO] Running 'npm install --no-optional' in /home/mgoodrich/openmrs/modules/coreapps
[INFO] 
[INFO] > canvas@1.1.6 install /home/mgoodrich/openmrs/modules/coreapps/node_modules/canvas
[INFO] > node-gyp rebuild
[INFO] 
[ERROR] sh: 1: node-gyp: Permission denied
[WARNING] npm WARN optional SKIPPING OPTIONAL DEPENDENCY: fsevents@1.2.7 (node_modules/fsevents):
[WARNING] npm WARN notsup SKIPPING OPTIONAL DEPENDENCY: Unsupported platform for fsevents@1.2.7: wanted {"os":"darwin","arch":"any"} (current: {"os":"linux","arch":"x64"})
[WARNING] npm WARN optional SKIPPING OPTIONAL DEPENDENCY: canvas@1.1.6 (node_modules/canvas):
[WARNING] npm WARN optional SKIPPING OPTIONAL DEPENDENCY: canvas@1.1.6 install: `node-gyp rebuild`
[WARNING] npm WARN optional SKIPPING OPTIONAL DEPENDENCY: Exit status 126
[INFO] 
[INFO] --- frontend-maven-plugin:1.4:npm (npm run build) @ coreapps-omod ---
[INFO] Running 'npm run build' in /home/mgoodrich/openmrs/modules/coreapps
[INFO] 
[INFO] > angularjs-coreapps@1.0.0 build /home/mgoodrich/openmrs/modules/coreapps
[INFO] > rimraf %npm_package_config_targetDir% && rimraf $npm_package_config_targetDir && karma start && webpack --progress --colors --env=prod
[INFO] 
[INFO] webpack: wait until bundle finished: 
[INFO] 
[INFO] webpack: Compiled successfully.
[INFO] 05 03 2021 15:07:00.487:INFO [karma]: Karma v1.7.1 server started at http://0.0.0.0:9876/
[INFO] 05 03 2021 15:07:00.488:INFO [launcher]: Launching browser PhantomJS with unlimited concurrency
[INFO] 05 03 2021 15:07:00.520:INFO [launcher]: Starting browser PhantomJS
[INFO] 05 03 2021 15:07:01.414:INFO [PhantomJS 2.1.1 (Linux 0.0.0)]: Connected on socket RGM8rOIIkz5ANx52AAAA with id 88245501
       PhantomJS 2.1.1 (Linux 0.0.0): Executed 2 of 2 SUCCESS (0.049 secs / 0.054 secs)

So I would think that removing that line means it wouldn't run the tests...

Comment (mogoodrich) [reply package.json:76]: What you probably could do is add another step to the Maven build, or change this one that does a "test and build":
https://github.com/openmrs/openmrs-module-coreapps/blob/master/omod/pom.xml#L395

Comment (mogoodrich) [reply package.json:76]: Here's a summary of the Maven Lifecycle (search for "default lifecycle" here):

https://maven.apache.org/guides/introduction/introduction-to-the-lifecycle.html

It looks like the the current plugin is set to plug in at the "generate-resources" stage of the build:

https://github.com/openmrs/openmrs-module-coreapps/blob/master/omod/pom.xml#L377

You could argue whether that's the best place or not, but if you don't want to really fool around with the maven config, you could just make sure a npm run test happens before a npm run build...

Comment (brandones) [reply package.json:76]: I've added `npm run test` to the mvn `test` phase. That phase runs after `generate-resources` and before `install`.

Comment (mogoodrich) [reply package.json:76]: Sounds good!

**Pull request merged**